pkg/experiment/metastore/dlq/recovery.go

```diff
@@ -0,0 +1,152 @@
+package dlq
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+	"github.com/grafana/pyroscope/pkg/tenant"
+
+	objstore "github.com/thanos-io/objstore"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type RecoveryConfig struct {
+	Period time.Duration
+}
+
+type LocalServer interface {
+	AddBlock(context.Context, *metastorev1.AddBlockRequest) (*metastorev1.AddBlockResponse, error)
+}
+
+type Recovery struct {
+	cfg    RecoveryConfig
+	l      log.Logger
+	srv    LocalServer
+	bucket objstore.Bucket
+
+	started bool
+	wg      sync.WaitGroup
+	m       sync.Mutex
+	cancel  func()
+}
+
+func NewRecovery(cfg RecoveryConfig, l log.Logger, srv LocalServer, bucket objstore.Bucket) *Recovery {
+	return &Recovery{
+		cfg:    cfg,
+		l:      l,
+		srv:    srv,
+		bucket: bucket,
+	}
+}
+
+func (r *Recovery) Start() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if r.started {
+		r.l.Log("msg", "recovery already started")
+		return
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.started = true
+	go r.recoverLoop(ctx)
+	r.l.Log("msg", "recovery started")
+}
+
+func (r *Recovery) Stop() {
+	r.m.Lock()
+	defer r.m.Unlock()
+	if !r.started {
+		r.l.Log("msg", "recovery already stopped")
+		return
+	}
+	r.cancel()
+	r.wg.Wait()
+	r.started = false
+	r.l.Log("msg", "recovery stopped")
+}
+
+const pathAnon = tenant.DefaultTenantID
+const pathBlock = "block.bin"
+const pathMetaPB = "meta.pb"
+const pathDLQ = "dlq"
+
+func (r *Recovery) recoverLoop(ctx context.Context) {
+	ticker := time.NewTicker(r.cfg.Period)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			r.recoverTick(ctx)
+		}
+	}
+}
+
+func (r *Recovery) recoverTick(ctx context.Context) {
+	err := r.bucket.Iter(ctx, pathDLQ, func(metaPath string) error {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		r.recover(ctx, metaPath)
+		return nil
+	}, objstore.WithRecursiveIter)
+	if err != nil {
+		level.Error(r.l).Log("msg", "failed to iterate over dlq", "err", err)
+	}
+}
```

Comment:
NB: One danger I see is that we mix old and new blocks in compaction jobs, which may increase the read amplification factor. I think we need to implement some logic to prevent this – just one more rule for the compaction planner. It’s not really critical, as we don't expect this to happen frequently.

/cc @aleks-p

Reply:
I'll add s little more context here.

Basically, when we try to add the same md entry twice there are three cases:
1. We still hold the entry in memory in the metastore. That's simple: we just reject it.
2. We've already compacted the block and we still have tombstone for it. That's simple: we just reject it.
3. We've already compacted the block and removed it from object storage and don't have tombstone for it. Somewhat tricky – I propose the following:
   * Handle the md entry as usual: add it to the index, add it to the compaction job, create a tombstone afterwards, etc. The entry is visible to readers.
   * Ignore "object not found" in compactors and queriers (reflect it in logs and metrics). Clearly, even if the object is missing due to any other reason, it's probably the best strategy. Better than failing the whole query/job at last.
   * Ensure that we only remove tombstone if the object is actually deleted from object storage.